tracing-subscriber/src/registry/mod.rs

```diff
@@ -45,3 +55,98 @@ pub trait LookupMetadata {
         self.metadata(id).is_some()
     }
 }
+
+pub trait LookupSpan<'a> {
+    type Data: SpanData<'a>;
+    fn span_data(&'a self, id: &Id) -> Option<Self::Data>;
+
+    fn span(&'a self, id: &Id) -> Option<SpanRef<'_, Self>>
+    where
+        Self: Sized,
+    {
+        let data = self.span_data(id)?;
+        Some(SpanRef {
+            registry: self,
+            data,
+        })
+    }
+}
+
+pub trait SpanData<'a> {
+    type Children: Iterator<Item = &'a Id>;
+    type Follows: Iterator<Item = &'a Id>;
+
+    fn id(&self) -> Id;
+    fn metadata(&self) -> &'static Metadata<'static>;
+    fn parent(&self) -> Option<&Id>;
+    fn children(&'a self) -> Self::Children;
+    fn follows_from(&'a self) -> Self::Follows;
+    fn extensions(&self) -> RwLockReadGuard<'_, Extensions>;
+    fn extensions_mut(&self) -> RwLockWriteGuard<'_, Extensions>;
```

Comment:
we should change the `RwLockReadGuard/RwLockWriteGuard`s here to structs we define, with private constructors (for now). Something like:

```rust
pub struct Extensions<'a> {
    // where `ExtensionsInner` is the type that used to be called `Extensions`, which can be made private
    inner: RwLockReadGuard<'a, ExtensionsInner>
}

pub struct ExtensionsMut<'a> {
    inner: RwLockWriteGuard<'a, ExtensionsInner>,
}
``` 
and then have impls of the extensions methods for those types. Alternatively, we could keep the public Extensions type as-is, and have the guards impl `Deref`/`DerefMut`, but then we'd need to call them, like, `ExtensionsRef` or something...

the first one, although boilerplate-ier, is probably better for forward compat...

Reply:
Opted for the first one, but I gotta give it better documentation at some point in the future.